Sources/SwiftProtobuf/TextDecoder.swift

```diff
@@ -21,67 +21,449 @@ import Swift
 /// pushback and convenience functions for iterating over complex
 /// structures.
 ///
-public struct TextDecoder: FieldDecoder {
-    private var scanner: TextScanner
-    public var complete: Bool {return scanner.complete}
-    public var rejectConflictingOneof: Bool {return true}
+public struct TextDecoder: Decoder {
+    internal var scanner: TextScanner
+    private var fieldCount = 0
+    private var terminator: UInt8?
+    private var fieldNameMap: FieldNameMap?
+    private var messageType: Message.Type?
 
-    internal init(text: String, extensions: ExtensionSet? = nil) {
+    internal var complete: Bool {
+        mutating get {
+            return scanner.complete
+        }
+    }
+
+    internal init(messageType: Message.Type, text: String, extensions: ExtensionSet?) throws {
         scanner = TextScanner(text: text, extensions: extensions)
+        guard let nameProviding = (messageType as? ProtoNameProviding.Type) else {
+            throw TextDecodingError.missingFieldNames
+        }
+        fieldNameMap = nameProviding._protobuf_fieldNames
+        self.messageType = messageType
     }
 
-    internal init(scanner: TextScanner) {
+    internal init(messageType: Message.Type, scanner: TextScanner, terminator: UInt8?) throws {
         self.scanner = scanner
+        self.terminator = terminator
+        guard let nameProviding = (messageType as? ProtoNameProviding.Type) else {
+            throw TextDecodingError.missingFieldNames
+        }
+        fieldNameMap = nameProviding._protobuf_fieldNames
+        self.messageType = messageType
     }
 
-    internal mutating func decodeFullObject<M: Message>(message: inout M, terminator: UInt8?) throws {
-        guard let nameProviding = (M.self as? ProtoNameProviding.Type) else {
-            throw DecodingError.missingFieldNames
+
+    public mutating func handleConflictingOneOf() throws {
+        throw TextDecodingError.conflictingOneOf
+    }
+
+    public mutating func nextFieldNumber() throws -> Int? {
```

Comment:
`nextFieldNumber()` is called from `decodeMessage()` to iterate over the field numbers.  This requires the decoder to store the field name map and other message-level parsing state.